internal/cloud/azure/application_identity.go

```diff
@@ -3,121 +3,149 @@ package azure
-	// The ID is the service principal ID, so technically theres a chance that last time the App was made
-	// but the SP was not. We need to check if the Application was made and skip creation if it was.
-	if config.ApplicationID == "" {
-		app, _, appErr := appClient.Create(ctx, msgraph.Application{
-			DisplayName: &config.Name,
-		})
-		if appErr != nil {
-			return appErr
-		}
-		config.ApplicationID = *app.ID
+	client, errClient := armmsi.NewUserAssignedIdentitiesClient(config.SubscriptionID.Value, cred, nil)
+	if errClient != nil {
+		return nil, errClient
 	}
 
-	// fetch the application to make sure it exists
-	app, _, err := appClient.Get(ctx, config.ApplicationID, odata.Query{})
+	return client, nil
+}
+
+func newFederatedIdentityCredentialClientFactory(ctx context.Context, config *AzureProviderConfig) (FederatedIdentityCredentialClient, error) {
+	cred, err := azidentity.NewClientSecretCredential(config.TenantID.Value, config.ClientID.Value, config.ClientSecret.Value, nil)
 	if err != nil {
-		return fmt.Errorf("error retrieving Application with object ID %v: %w", config.ApplicationID, err)
+		return nil, err
 	}
 
-	sp, _, spErr := spClient.Create(ctx, msgraph.ServicePrincipal{
-		AppId: app.AppId,
-	})
-	if spErr != nil {
-		return spErr
-	}
-	config.ServicePrincipalID = *sp.ID
-	config.ServicePrincipalClientID = *sp.AppId
-	config.ID = *sp.ID
-
-	// fetch the service principal to make sure it exists
-	_, _, spCheckErr := spClient.Get(ctx, config.ServicePrincipalID, odata.Query{})
-	if spCheckErr != nil {
-		return fmt.Errorf("error retrieving Service Principal with ID %v: %w", config.ServicePrincipalID, spCheckErr)
+	client, errClient := armmsi.NewFederatedIdentityCredentialsClient(config.SubscriptionID.Value, cred, nil)
+	if errClient != nil {
+		return nil, errClient
 	}
 
-	// Technically the secret is only needed for Kubernetes until the support workload identity. Maybe we make this on a conditional?
-	spSecret, _, secretErr := spClient.AddPassword(ctx, *sp.ID, msgraph.PasswordCredential{})
-	if secretErr != nil {
-		return secretErr
+	return client, nil
+}
+
+func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityConfig, client ManagedIdentityClient, fedClient FederatedIdentityCredentialClient) error {
+	identity, errCreate := client.CreateOrUpdate(ctx,
+		config.ResourceGroupName,
+		config.Name,
+		armmsi.Identity{
+			Location: &config.Location,
+			Tags: map[string]*string{
+				"managed-by": to.Ptr("massdriver"),
+			},
+		},
+		nil,
+	)
+	if errCreate != nil {
+		return errCreate
 	}
-	config.ServicePrincipalSecret = *spSecret.SecretText
 
+	resourceID := *identity.ID
+	id := strings.Replace(resourceID, "resourcegroup", "resourceGroup", -1)
+	config.ID = *identity.Properties.PrincipalID
+	config.ClientID = *identity.Properties.ClientID
+	config.TenantID = *identity.Properties.TenantID
+	config.ResourceID = id
+
+	if config.KubernetesNamspace != "" {
```

Comment:
This seems like a bit of a fragile check, but probably fine for now. It seems like there are special things we need to do in each cloud if its Kubernetes. In AWS we need a very specific `AssumeRole` permission. Not sure about GKE. Once this is stable we may need to do a sweep over the resource parameters to see of there are optimizations, particularly around Kubernetes.

Reply:
I agree. I think it's what we do for the k8s stuff in GCP, would love to improve both of them.